pallets/pool-fees/src/lib.rs

```diff
@@ -0,0 +1,590 @@
+// Copyright 2023 Centrifuge Foundation (centrifuge.io).
+
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+pub mod types;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use cfg_traits::{
+		changes::ChangeGuard,
+		fee::{FeeAmountProration, PoolFees},
+		investments::TrancheCurrency,
+		Permissions, PoolInspect, PoolReserve, SaturatedProration, TimeAsSecs,
+	};
+	use cfg_types::{
+		permissions::{PermissionScope, PoolRole, Role},
+		pools::{FeeAmount, FeeAmountType, FeeAmountType::Fixed, FeeBucket, PoolFee},
+	};
+	use codec::HasCompact;
+	use frame_support::{
+		pallet_prelude::*,
+		traits::fungibles::{Inspect, Mutate},
+		weights::Weight,
+	};
+	use frame_system::pallet_prelude::*;
+	use sp_arithmetic::{
+		traits::{
+			AtLeast32BitUnsigned, EnsureAdd, EnsureAddAssign, EnsureSubAssign, One, Saturating,
+			Zero,
+		},
+		ArithmeticError, FixedPointOperand,
+	};
+	use sp_std::vec::Vec;
+
+	use super::*;
+	use crate::types::{Change, DisbursingFee};
+
+	pub type PoolFeeOf<T> = PoolFee<
+		<T as frame_system::Config>::AccountId,
+		<T as Config>::Balance,
+		<T as Config>::Rate,
+	>;
+
+	pub type DisbursingFeeOf<T> =
+		DisbursingFee<<T as frame_system::Config>::AccountId, <T as Config>::Balance>;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The identifier of a particular fee
+		type FeeId: Parameter
+			+ Member
+			+ Default
+			+ TypeInfo
+			+ MaxEncodedLen
+			+ Copy
+			+ EnsureAdd
+			+ One
+			+ Ord;
+
+		/// The source of truth for the balance of accounts
+		type Balance: Parameter
+			+ Member
+			+ AtLeast32BitUnsigned
+			+ Default
+			+ Copy
+			+ FixedPointOperand
+			+ SaturatedProration<Time = Self::Time>
+			+ From<Self::Time>
+			+ MaybeSerializeDeserialize
+			+ MaxEncodedLen;
+
+		/// The currency type of transferrable tokens
+		type CurrencyId: Parameter + Member + Copy + TypeInfo + MaxEncodedLen;
+
+		/// The pool id type required for the investment identifier
+		type PoolId: Member
+			+ Parameter
+			+ Default
+			+ Copy
+			+ HasCompact
+			+ MaxEncodedLen
+			+ Into<
```

Comment:
Unfortunately it is. Otherwise I need to add `where` bounds:

```
the trait bound `<<T as pallet::Config>::PoolReserve as PoolInspect<<T as frame_system::Config>::AccountId, <T as pallet::Config>::CurrencyId>>::PoolId: From<<T as pallet::Config>::PoolId>` is not satisfied
```

Reply:
Then I think in your `T::PoolReserve`, you can add an extra bound indicating it really is the same. Something like:
```rust
T::PoolReserve: T::PoolReserve<> + T::PoolInspect<PoolId = Self::PoolId>
```

One question related to this. Why are there two associated types `T::PoolInspect` and `T::PoolReserve` if `PoolReserve` is also a `PoolInspect`?

Reply:
Great point. `PoolReserve` is sufficient. Also the restriction of the associated `PoolId` type make the above trait restriction redundant.